File: src/jvm/package/src/doc-examples/java/questpdf/docexamples/text/TextBasicExamples.java
package questpdf.docexamples.text;

import com.questpdf.fluent.Document;
import com.questpdf.helpers.Colors;
import com.questpdf.helpers.PageSize;
import com.questpdf.helpers.PageSizes;
import com.questpdf.helpers.Placeholders;
import com.questpdf.infrastructure.IContainer;
import com.questpdf.infrastructure.ImageCompressionQuality;
import com.questpdf.infrastructure.ImageFormat;
import com.questpdf.infrastructure.ImageGenerationSettings;
import com.questpdf.infrastructure.TextStyle;
import org.junit.jupiter.api.Test;
import questpdf.docexamples.DocExample;

import java.util.function.Function;

public class TextBasicExamples extends DocExample {

    @Test
    public void basic() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text("Sample text");
                });
            })
            .generateImages(index -> output("text-basic.webp"), settings);
    }

    @Test
    public void basicWithStyle() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .column(column -> {
                            column.spacing(10f);

                            // The C# original routes CellStyle through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; a local Function is the direct equivalent.
                            Function<IContainer, IContainer> cellStyle = container ->
                                container.background(Colors.Grey.getLighten3()).padding(10f);

                            cellStyle.apply(column.item())
                                .text("Text with blue color")
                                .fontColor(Colors.Blue.getDarken1());

                            cellStyle.apply(column.item())
                                .text("Bold and underlined text")
                                .bold()
                                .underline();

                            cellStyle.apply(column.item())
                                .text("Centered small text")
                                .fontSize(12f)
                                .alignCenter();
                        });
                });
            })
            .generateImages(index -> output("text-basic-descriptor.webp"), settings);
    }

    @Test
    public void rich() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(500f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.alignCenter();

                            text.span("The ");
                            text.span("chemical formula").underline();
                            text.span(" of ");
                            text.span("sulfuric acid").backgroundColor(Colors.Amber.getLighten3());
                            text.span(" is H");
                            text.span("2").subscript();
                            text.span("SO");
                            text.span("4").subscript();
                            text.span(".");
                        });
                });
            })
            .generateImages(index -> output("text-rich.webp"), settings);
    }

    @Test
    public void styleInheritance() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.minSize(new PageSize(0f, 0f));
                    page.maxSize(new PageSize(600f, 1000f));
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .defaultTextStyle(style -> style.fontSize(20f))
                        .column(column -> {
                            column.spacing(10f);

                            column.item().text("Products").extraBold().underline().decorationThickness(2f);

                            column.item().text("Comments: " + Placeholders.sentence());

                            column.item()
                                .defaultTextStyle(style -> style.fontSize(14f))
                                .table(table -> {
                                    table.columnsDefinition(columns -> {
                                        columns.constantColumn(30f);
                                        columns.relativeColumn(1f);
                                        columns.relativeColumn(2f);
                                    });

                                    table.header(header -> {
                                        // The C# original routes Style through the Element(Func<IContainer, IContainer>)
                                        // overload, which is not bridged; a local Function is the direct equivalent.
                                        Function<IContainer, IContainer> style = container ->
                                            container.background(Colors.Grey.getLighten3())
                                                .borderBottom(1f)
                                                .paddingHorizontal(5f)
                                                .paddingVertical(10f)
                                                .defaultTextStyle(textStyle -> textStyle.bold().fontColor(Colors.Blue.getMedium()));

                                        style.apply(header.cell()).text("ID");
                                        style.apply(header.cell()).text("Name");
                                        style.apply(header.cell()).text("Description");
                                    });

                                    // The C# original routes Style through the Element(Func<IContainer, IContainer>)
                                    // overload, which is not bridged; a local Function is the direct equivalent.
                                    Function<IContainer, IContainer> style = container ->
                                        container.padding(5f);

                                    for (var i = 0; i < 5; i++) {
                                        style.apply(table.cell()).text(String.valueOf(i)).bold();
                                        style.apply(table.cell()).text(Placeholders.label());
                                        style.apply(table.cell()).text(Placeholders.sentence());
                                    }
                                });
                        });
                });
            })
            .generateImages(index -> output("text-style-inheritance.webp"), settings);
    }

    @Test
    public void pageNumber() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.size(PageSizes.getA5());
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .extend()
                        .placeholder();

                    page.footer()
                        .paddingTop(25f)
                        .alignCenter()
                        .text("3 / 10");
                        // .text(text -> {
                        //     text.currentPageNumber();
                        //     text.span(" / ");
                        //     text.totalPages();
                        // });
                });
            })
            .generateImages(index -> output("text-page-number.webp"), settings);
    }

    @Test
    public void pageNumberFormat() {
        var settings = new ImageGenerationSettings();
        settings.setImageFormat(ImageFormat.Webp);
        settings.setImageCompressionQuality(ImageCompressionQuality.VeryHigh);
        settings.setRasterDpi(144);

        Document
            .create(document -> {
                document.page(page -> {
                    page.size(PageSizes.getA5());
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            text.currentPageNumber().format(TextBasicExamples::formatWithLeadingZeros);
                        });
                });
            })
            .generateImages(index -> output("text-page-number-format.webp"), settings);
    }

    private static String formatWithLeadingZeros(Integer pageNumber) {
        var expectedLength = 3;
        var number = pageNumber != null ? pageNumber : 1;
        var text = Integer.toString(number);
        return "0".repeat(Math.max(expectedLength - text.length(), 0)) + text;
    }

    @Test
    public void hyperlink() {
        Document
            .create(document -> {
                document.page(page -> {
                    page.size(PageSizes.getA6().landscape());
                    page.defaultTextStyle(style -> style.fontSize(20f));
                    page.margin(25f);

                    page.content()
                        .text(text -> {
                            var hyperlinkStyle = TextStyle.getDefault()
                                .fontColor(Colors.Blue.getMedium())
                                .underline();

                            text.span("To learn more about QuestPDF, please visit its ");
                            text.hyperlink("homepage", "https://www.questpdf.com/").style(hyperlinkStyle);
                            text.span(", ");
                            text.hyperlink("GitHub repository", "https://github.com/QuestPDF/QuestPDF").style(hyperlinkStyle);
                            text.span(" and ");
                            text.hyperlink("NuGet package page", "https://www.nuget.org/packages/QuestPDF").style(hyperlinkStyle);
                            text.span(".");
                        });
                });
            })
            .generatePdf(output("text-hyperlink.pdf"));
    }
}

File: src/jvm/package/src/doc-examples/kotlin/questpdf/docexamples/text/TextBasicExamples.kt
package questpdf.docexamples.text

import com.questpdf.fluent.Document
import com.questpdf.helpers.Colors
import com.questpdf.helpers.PageSize
import com.questpdf.helpers.PageSizes
import com.questpdf.helpers.Placeholders
import com.questpdf.infrastructure.IContainer
import com.questpdf.infrastructure.ImageCompressionQuality
import com.questpdf.infrastructure.ImageFormat
import com.questpdf.infrastructure.ImageGenerationSettings
import com.questpdf.infrastructure.TextStyle
import org.junit.jupiter.api.Test
import questpdf.docexamples.DocExample

class TextBasicExamples : DocExample() {

    @Test
    fun basic() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text("Sample text")
                }
            }
            .generateImages({ output("text-basic.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun basicWithStyle() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .column {
                            spacing(10f)

                            // The C# original routes CellStyle through the Element(Func<IContainer, IContainer>)
                            // overload, which is not bridged; an extension function is the direct equivalent.
                            fun IContainer.cellStyle(): IContainer =
                                background(Colors.Grey.Lighten3).padding(10f)

                            item()
                                .cellStyle()
                                .text("Text with blue color")
                                .fontColor(Colors.Blue.Darken1)

                            item()
                                .cellStyle()
                                .text("Bold and underlined text")
                                .bold()
                                .underline()

                            item()
                                .cellStyle()
                                .text("Centered small text")
                                .fontSize(12f)
                                .alignCenter()
                        }
                }
            }
            .generateImages({ output("text-basic-descriptor.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun rich() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(500f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            alignCenter()

                            span("The ")
                            span("chemical formula").underline()
                            span(" of ")
                            span("sulfuric acid").backgroundColor(Colors.Amber.Lighten3)
                            span(" is H")
                            span("2").subscript()
                            span("SO")
                            span("4").subscript()
                            span(".")
                        }
                }
            }
            .generateImages({ output("text-rich.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun styleInheritance() {
        Document
            .create {
                page {
                    minSize(PageSize(0f, 0f))
                    maxSize(PageSize(600f, 1000f))
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .defaultTextStyle { fontSize(20f) }
                        .column {
                            spacing(10f)

                            item().text("Products").extraBold().underline().decorationThickness(2f)

                            item().text("Comments: " + Placeholders.sentence())

                            item()
                                .defaultTextStyle { fontSize(14f) }
                                .table {
                                    columnsDefinition {
                                        constantColumn(30f)
                                        relativeColumn(1f)
                                        relativeColumn(2f)
                                    }

                                    header {
                                        // The C# original routes Style through the Element(Func<IContainer, IContainer>)
                                        // overload, which is not bridged; an extension function is the direct equivalent.
                                        fun IContainer.style(): IContainer =
                                            background(Colors.Grey.Lighten3)
                                                .borderBottom(1f)
                                                .paddingHorizontal(5f)
                                                .paddingVertical(10f)
                                                .defaultTextStyle { bold().fontColor(Colors.Blue.Medium) }

                                        cell().style().text("ID")
                                        cell().style().text("Name")
                                        cell().style().text("Description")
                                    }

                                    // The C# original routes Style through the Element(Func<IContainer, IContainer>)
                                    // overload, which is not bridged; an extension function is the direct equivalent.
                                    fun IContainer.style(): IContainer =
                                        padding(5f)

                                    for (i in 0 until 5) {
                                        cell().style().text(i.toString()).bold()
                                        cell().style().text(Placeholders.label())
                                        cell().style().text(Placeholders.sentence())
                                    }
                                }
                        }
                }
            }
            .generateImages({ output("text-style-inheritance.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun pageNumber() {
        Document
            .create {
                page {
                    size(PageSizes.A5)
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .extend()
                        .placeholder()

                    footer()
                        .paddingTop(25f)
                        .alignCenter()
                        .text("3 / 10")
                        // .text {
                        //     currentPageNumber()
                        //     span(" / ")
                        //     totalPages()
                        // }
                }
            }
            .generateImages({ output("text-page-number.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun pageNumberFormat() {
        Document
            .create {
                page {
                    size(PageSizes.A5)
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    fun formatWithLeadingZeros(pageNumber: Int?): String {
                        val expectedLength = 3
                        val number = pageNumber ?: 1
                        return number.toString().padStart(expectedLength, '0')
                    }

                    content()
                        .text {
                            currentPageNumber().format(::formatWithLeadingZeros)
                        }
                }
            }
            .generateImages({ output("text-page-number-format.webp") }, ImageGenerationSettings().apply {
                imageFormat = ImageFormat.Webp
                imageCompressionQuality = ImageCompressionQuality.VeryHigh
                rasterDpi = 144
            })
    }

    @Test
    fun hyperlink() {
        Document
            .create {
                page {
                    size(PageSizes.A6.landscape())
                    defaultTextStyle { fontSize(20f) }
                    margin(25f)

                    content()
                        .text {
                            val hyperlinkStyle = TextStyle.Default
                                .fontColor(Colors.Blue.Medium)
                                .underline()

                            span("To learn more about QuestPDF, please visit its ")
                            hyperlink("homepage", "https://www.questpdf.com/").style(hyperlinkStyle)
                            span(", ")
                            hyperlink("GitHub repository", "https://github.com/QuestPDF/QuestPDF").style(hyperlinkStyle)
                            span(" and ")
                            hyperlink("NuGet package page", "https://www.nuget.org/packages/QuestPDF").style(hyperlinkStyle)
                            span(".")
                        }
                }
            }
            .generatePdf(output("text-hyperlink.pdf"))
    }
}
